be/src/storage/rowset/vectorized/segment_iterator.cpp

```diff
@@ -557,27 +575,48 @@ Status SegmentIterator::_read_columns(const Schema& schema, Chunk* chunk, size_t
 }
 
 inline Status SegmentIterator::_read(Chunk* chunk, vector<rowid_t>* rowid, size_t n) {
-    Range r = _range_iter.next(n);
-    size_t nread = r.span_size();
-    if (_cur_rowid != r.begin() || _cur_rowid == 0) {
-        _cur_rowid = r.begin();
+    size_t read_num = 0;
+    SparseRange range;
+    std::vector<rowid_t> read_rowid;
+    read_rowid.reserve(n);
+    size_t chunk_rowid_start = _cur_rowid;
+    size_t cur_rowid = _cur_rowid;
+
+    if (_cur_rowid != _range_iter.begin() || _cur_rowid == 0) {
+        _cur_rowid = _range_iter.begin();
         _opts.stats->block_seek_num += 1;
         SCOPED_RAW_TIMER(&_opts.stats->block_seek_ns);
         RETURN_IF_ERROR(_context->seek_columns(_cur_rowid));
     }
+
+    while ((read_num < n) && _range_iter.has_more()) {
+        Range r = _range_iter.next(n - read_num);
+        read_num += r.span_size();
+        chunk_rowid_start = r.begin();
+        cur_rowid = r.end();
+        range.add(r);
+    }
+
     {
         _opts.stats->blocks_load += 1;
         SCOPED_RAW_TIMER(&_opts.stats->block_fetch_ns);
-        RETURN_IF_ERROR(_context->read_columns(chunk, nread));
+        RETURN_IF_ERROR(_context->read_columns(chunk, range));
     }
-    _chunk_rowid_start = _cur_rowid;
+
     if (rowid != nullptr) {
-        for (uint32_t i = _cur_rowid; i < _cur_rowid + nread; i++) {
-            rowid->push_back(i);
+        rowid->reserve(range.span_size() + n);
```

Comment:
it seems that range.span_size == n, I think you want to reserve rowid->size() + n?
And I advise to rename rowid to rowids.

Reply:
done